common/fs/src/tail.rs

```diff
@@ -87,6 +88,199 @@ impl Tailer {
         None
     }
 
+    async fn get_initial_offset(
+        target: &Path,
+        fs: &FileSystem,
+        initial_offsets: Option<HashMap<FileId, u64>>,
+        lookback_config: Lookback,
+    ) -> Option<(EntryKey, u64)> {
+        let entry_key = fs.lookup(target, &fs.entries.borrow())?;
+        let entries = fs.entries.borrow();
+        let entry = &entries.get(entry_key)?;
+        let path = fs.resolve_direct_path(entry, &fs.entries.borrow());
+        if let Entry::File { data, .. } = entry {
+            let inode: FileId = { (&data.borrow().deref().get_inode().await).into() };
+            Some((
+                entry_key,
+                match lookback_config {
+                    Lookback::Start => match initial_offsets.as_ref() {
+                        Some(initial_offsets) => {
+                            let offset = initial_offsets.get(&inode).copied().unwrap_or(0);
+                            debug!("Got offset {} from state using key {:?}", offset, path);
+                            offset
+                        }
+                        None => 0,
+                    },
+                    Lookback::SmallFiles => {
+                        match initial_offsets.as_ref() {
+                            Some(initial_offsets) => {
+                                let offset = initial_offsets.get(&inode).copied().unwrap_or(0);
+                                debug!("Got offset {} from state using key {:?}", offset, path);
+                                offset
+                            }
+                            None => {
+                                // Check the actual file len
+                                let len = path.metadata().map(|m| m.len()).unwrap_or(0);
+                                debug!("Smallfiles lookback {} from len using key {:?}", len, path);
+                                if len < 8192 {
+                                    0
+                                } else {
+                                    len
+                                }
+                            }
+                        }
+                    }
+                    Lookback::None => path.metadata().map(|m| m.len()).unwrap_or(0),
+                },
+            ))
+        } else {
+            None
+        }
+    }
+
+    async fn handle_event(
+        event: Event,
+        initial_offsets: Option<HashMap<FileId, u64>>,
+        lookback_config: Lookback,
+        fs: &FileSystem,
+    ) -> Option<impl Stream<Item = LazyLineSerializer>> {
+        match event {
+            Event::Initialize(entry_ptr) => {
+                debug!("Initialise Event");
```

Comment:
We should continue using the same messages that we defined for 3.1 "`initialize`".